source/common/network/apple_dns_impl.cc

```diff
@@ -0,0 +1,259 @@
+    throw EnvoyException("Error in DNSServiceRefSockFD");
+  }
+  ENVOY_LOG(warn, "DNS resolver has fd={}", fd);
+
+  sd_ref_event_ = dispatcher_.createFileEvent(
+      fd, [this](uint32_t events) { onEventCallback(events); }, Event::FileTriggerType::Level,
+      Event::FileReadyType::Read);
+  sd_ref_event_->setEnabled(Event::FileReadyType::Read);
+}
+
+void AppleDnsResolverImpl::onEventCallback(uint32_t events) {
+  ENVOY_LOG(warn, "DNS resolver file event");
+  if (events & Event::FileReadyType::Read) {
+    DNSServiceProcessResult(main_sd_ref_);
+  }
+}
+
+ActiveDnsQuery* AppleDnsResolverImpl::resolve(const std::string& dns_name,
+                                              DnsLookupFamily dns_lookup_family,
+                                              ResolveCb callback) {
+  ENVOY_LOG(warn, "DNS resolver resolve={}", dns_name);
+  std::unique_ptr<PendingResolution> pending_resolution(
+      new PendingResolution(*this, callback, dispatcher_, main_sd_ref_, dns_name));
+
+  DNSServiceErrorType error = pending_resolution->dnsServiceGetAddrInfo(dns_lookup_family);
+  if (error != kDNSServiceErr_NoError) {
+    return nullptr;
+  }
+
+  // If the query was synchronously resolved, there is no need to return the query.
+  // FIXME: I don't think synchronous resolution is possible from my read of dns_sd.h,
+  // and tests with local resolution.
+  // What about with caching?
+  if (pending_resolution->synchronously_completed_) {
+    return nullptr;
+  }
+
+  pending_resolution->owned_ = true;
+  return pending_resolution.release();
+}
+
+void AppleDnsResolverImpl::flushPendingQueries() {
+  ENVOY_LOG(warn, "DNS Resolver flushing {} queries", queries_with_pending_cb_.size());
+  for (std::list<PendingResolution*>::iterator it = queries_with_pending_cb_.begin();
+       it != queries_with_pending_cb_.end(); ++it) {
+    try {
+      ASSERT((*it)->pending_cb_);
+      (*it)->callback_((*it)->pending_cb_->status_, std::move((*it)->pending_cb_->responses_));
+    } catch (const EnvoyException& e) {
+      ENVOY_LOG(critical, "EnvoyException in DNSService callback: {}", e.what());
+      (*it)->dispatcher_.post([s = std::string(e.what())] { throw EnvoyException(s); });
+    } catch (const std::exception& e) {
+      ENVOY_LOG(critical, "std::exception in DNSService callback: {}", e.what());
+      (*it)->dispatcher_.post([s = std::string(e.what())] { throw EnvoyException(s); });
+    } catch (...) {
+      ENVOY_LOG(critical, "Unknown exception in DNSService callback");
+      (*it)->dispatcher_.post([] { throw EnvoyException("unknown"); });
+    }
+
+    if ((*it)->owned_) {
+      ENVOY_LOG(warn, "Resolution for {} completed (async)", (*it)->dns_name_);
+      delete *it;
+    } else {
+      ENVOY_LOG(warn, "Resolution for {} completed (synchronously)", (*it)->dns_name_);
+      (*it)->synchronously_completed_ = true;
+    }
+  }
+
+  // Purge the contents so no one tries to delete them again.
+  queries_with_pending_cb_.clear();
```

Comment:
Is this logic correct? Again I don't know the API, but it seems like you might be clearing all pending queries when you only get a response for some of them? Maybe more comments?

Reply:
There are two paths to flushing:
a) either an error happens
b) the MoreComing flag is not set
Because the connection to the DNS daemon is shared, a callback with either of those conditions should trigger a flush.

I commented in the header for the `queries_with_pending_cb_` list, but you are right, more comments are needed here; specially because the API has non-intuitive behavior.

Reply:
OK thanks. Per my other comment you might consider not doing the shared thing at first, but up to you.